### PR TITLE
[dask] test training when a worker has no data

### PR DESCRIPTION
**Pull request by jmoralez**
This includes a test to check that the training process succeeds even when a worker has no data, e.g. verify that #3882 is solved. 

I wanted to persist the collections to a specific worker with something like:
```python
for worker in workers:
    dX, dy, dw = client.persist([dX, dy, dw], workers=worker)
```
However I got: `TypeError: unhashable type: 'Array'`.

So in the test I just create a collection with one chunk, persist it and check that it is only in one worker. Any feedback to make this test more robust is welcome.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much! I have some ideas to simplify this, without needing to introduce a new function to check that the data is only on one worker.

1. Assert that `client.nworkers > 1`, just to be sure the client fixture is for a multi-worker cluster.
2. Just repartition the data to 1 partition (`.repartition(npartitions=1)` for data frame, `.rechunk((X.shape[0], X.shape[1]))`  for array). assert that `dX.npartitions == 1`. If you have only 1 partition, you know for sure only one worker has data!

**other requests**

Please make sure this test is run for ranking, classification, and regression. You can borrow from https://github.com/microsoft/LightGBM/blob/876bfe5f6c756e946594e3316fec716ae358de18/tests/python_package_test/test_dask.py#L463-L487 to see how to do that.

For training on a single worker, the result should be identical to non-Dask training. So please also train a regular `lightgbm.sklearn.LGBM[Classifier/Regressor/Ranker]` and then  check that `assert_eq(dask_model.predict(dX).compute(), local_model.predict(X))`. You can see the other tests in this file for references on how to do that, or `@` me if you have any questions.

**Comment by jmoralez** [comment]
Hi James. I've included the different tasks and outputs. I just saw the PR you made for the categoricals and saw the new output type, I'll wait til that gets merged to include it.

File: tests/python_package_test/test_dask.py
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:812]:
```suggestion
def test_training_succeeds_even_if_some_workers_do_not_have_any_daata(client, task, output):
```

I'd prefer a more informative name here.

File: tests/python_package_test/test_dask.py
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:814]:
```suggestion
    def repartition(collection):
        """Merge the parts of a Dask collection into a single partition"""
```

I love this function! But it has a very general name, so I'd like to recommend adding this docstring so others working on this code in the future will understand what its purpose is.

File: tests/python_package_test/test_dask.py
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:833]:
```suggestion
        g = None
        dg = None
```

Please put these on separate lines.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks so much! I like the changes you've made. We're starting to have enough of these tests where `tasks` is parameterized, makes sense to concentrate it at the top of `tetst_dask.py` like you did.

I left a few minor suggestions. Other than those, I agree with your proposal to wait until #3908 is merged.

**Comment by jmoralez** [comment]
Hi James. I've included your comments and the `dataframe-with-categorical` data output. Looking forward to your review.

**Review by jameslamb — APPROVED** [review]
This look great! Thanks very much for the help

**Pull request merged**
